rfcs/0014-group-by.md

```diff
@@ -0,0 +1,97 @@
+}
+```
+
+The intent is that the same row set should be grouped and aggregated if `groups` is specified.
+
+The `Grouping` structure looks like this:
+
+```rust
+pub struct Grouping {
+    /// Dimensions along which to partition the data
+    pub dimensions: IndexMap<String, Dimension>,
+    /// Aggregates to compute in each group
+    pub aggregates: IndexMap<String, Aggregate>,
+}
+```
+
+A `Grouping` slices a row set along several dimensions, and then performs some aggregations over each slice.
+
+A `Dimension` is an enum, but there is only one option right now, which is to slice by the value of a column:
+
+```rust
+pub enum Dimension {
+    Column {
+        /// The name of the column
+        column_name: String,
+        /// Path to a nested field within an object column
+        field_path: Option<Vec<String>>,
+    },
+}
+```
+
+We will support nested fields as dimensions.
+
+We might want to extend this enum later, for example with related collections, or multi-faceted dimensions (like timestamps faceted by various resolutions), or other kinds of dimensions.
+
+`Aggregate` is just the usual type of aggregates.
+
+The `RowSet` type is also extended to include `groups` when requested:
+
+```rust
+pub struct RowSet {
+    /// The results of the aggregates returned by the query
+    pub aggregates: Option<IndexMap<String, serde_json::Value>>,
+    /// The rows returned by the query, corresponding to the query's fields
+    pub rows: Option<Vec<IndexMap<String, RowFieldValue>>>,
+    /// The results of any grouping operation applies to the returned rows
+    pub groups: Option<Vec<Group>>,
+}
+```
+
+A `Group` is defined as follows:
+
+```rust
+pub struct Group {
+    /// Values of dimensions which identify this group
+    pub dimensions: IndexMap<String, serde_json::Value>,
+    /// Aggregates computed within this group
+    pub aggregates: IndexMap<String, serde_json::Value>,
+}
+```
+
+Requirements:
+
+- If `groups` is specified, then a `groups` property should also be provided in each row set returned.
+- The returned groups should correspond to a partition of the row set computed from the query:
+  - Within each partition, the chosen dimensions must be equal. These common values are hoisted out to the `dimensions` property of the `Group`.
+  - `aggregates` are computed over each partition in turn, in the same way as the aggregates API would compute them over the selected rows.
+
+The exact semantics of this partitioning operation are unspecified. For example, one connector might implement sort-then-group, and another might choose not to sort and return multiple groups for the same dimensions instead.
+
+## Future Work
+
+- Support ordering and filtering after the grouping operation
+  - Post-group filtering is called `HAVING` in SQL.
+
+## Open Questions
+
+- A `Grouping` with no dimensions is just a regular aggregation, do we want to unify the two?
+- Do we want to support multiple grouping operations in the same query?
```

Comment:
Can't the user just request groups twice, underneath two separate aliased fields? Presumably we'd want that to be run as a single query if we can.

Reply:
Two separate aliased root fields are still two separate queries to NDC right now.

Am I missing the point you're trying to make? Perhaps an example in GraphQL would help.

Reply:
Let's say I want to group the same rows two different ways. Then I need to make two queries, but a mutation might run in between them, which would make the combined result invalid.

On the other hand, I can fetch the same field twice with different arguments, say, in the same query.

Reply:
I guess my point is that this is fundamentally a problem with NDC as a whole right now, not just grouping. If I want to query the same table twice with different filters inside the same transaction, I can't. If I want to perform aggregations against the same table, but with a different input filter, inside the same transaction, I can't.

It seems to me that this could be solved more generally by allowing multiple query requests to be sent and executed inside the same transaction, rather than creating a special case for grouping, when from the graphql perspective all of these different query types that I may want to group together look the same (ie. a single graphql query with multiple root fields, aliased or otherwise).

Reply:
Yes, that's true. Not worth worrying about here, we can address this more generally later.